Utilities/StaticAnalyzers/src/MutableMemberAccessChecker.cpp

```diff
@@ -0,0 +1,116 @@
+//== MutableMemberAccessChecker.cpp - Checks for accessing mutable members via const pointer --------------*- C++ -*--==//
+//
+// By Ivan Razumov <ivan.razumov@cern.ch>
+//
+//===----------------------------------------------------------------------===//
+
+#include "MutableMemberAccessChecker.h"
+#include <clang/AST/Stmt.h>
+#include <clang/AST/Expr.h>
+#include <clang/AST/ExprCXX.h>
+#include <clang/AST/ParentMap.h>
+#include <clang/Analysis/AnalysisDeclContext.h>
+#include <clang/StaticAnalyzer/Core/PathSensitive/CheckerContext.h>
+
+namespace clangcms {
+  void MutableMemberAccessChecker::checkPreStmt(const clang::MemberExpr *ME, clang::ento::CheckerContext &C) const {
+    // == Only proceed if the member is mutable ==
+    const auto *FD = llvm::dyn_cast<clang::FieldDecl>(ME->getMemberDecl());
+    if (!FD || !FD->isMutable()) {
+      return;  // Skip if it's not a mutable field
+    }
+
+    // == Filter out classes with "safe" names ==
+    const auto *RD = llvm::dyn_cast<clang::CXXRecordDecl>(ME->getMemberDecl()->getDeclContext());
+    if (RD) {
+      std::string ClassName = RD->getNameAsString();
+      if (support::isSafeClassName(ClassName)) {
+        return;  // Skip checking for this class
+      }
+    }
+
+    // == Check if this is a modifying statement ==
+    bool isModification = false;
+
+    // Retrieve the parent statement of the MemberExpr
+    const clang::LocationContext *LC = C.getLocationContext();
+    const clang::ParentMap &PM = LC->getParentMap();
+    const clang::Stmt *ParentStmt = PM.getParent(ME);
+
+    if (!ParentStmt) {
+      return;
+    }
+
+    // Check if it is an assignment operator (binary operator)
+    if (const auto *BO = llvm::dyn_cast<clang::BinaryOperator>(ParentStmt)) {
+      if (BO->isAssignmentOp() && BO->getLHS() == ME) {
+        // The MemberExpr is on the left-hand side of an assignment
+        isModification = true;
+      }
+    }
+
+    // Check for increment/decrement
+    if (const auto *UO = llvm::dyn_cast<clang::UnaryOperator>(ParentStmt)) {
+      if (UO->isIncrementDecrementOp() && UO->getSubExpr() == ME) {
+        isModification = true;
+      }
+    }
+
+    if (!isModification) {
```

Comment:
Does calling a non-`const` member function count as "modification" here?

Reply:
One can't call non-`const` function from `const` one:

> in a member function with const qualifier, only other member functions with const qualifier may be called normally. 

[cppreference](https://en.cppreference.com/w/cpp/language/member_functions#Member_functions_with_cv-qualifiers)

Reply:
One can call a non-`const` member function of a data member that was declared as `mutable`.

Reply:
Thanks for catching. I have added a check for this case.